### PR TITLE
Laboratory work #3, Arina Savina - 22FPL1

### PR DESCRIPTION
**Pull request by a00rina**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or not text:
+            return None
+        tokens = []
+        alphas = 0
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:49]: too much... try anather way to chec

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -76,6 +97,11 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if element_id not in self._storage.values() or not isinstance(element_id, int):
+            return None
+        for key, value in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:102]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -159,6 +219,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        post_text = ''
+        for token, i in enumerate(decoded_corpus):
+            if token == self._end_of_word_token:
+                post_text += ' '
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:227]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -159,6 +219,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+            elif token == self._end_of_word_token and i == len(decoded_corpus) - 1:
+                post_text += '.'
+            else:
+                post_text += token
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:231]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py:102]: done

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py:49]: hoped that understood you right

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py]: done, hope it's right)

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -41,6 +44,33 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+                tokens.append(token)
+            elif token.isspace() and tokens[-1] != self._end_of_word_token:
+                tokens.append(self._end_of_word_token)
+        if (text[-1] in string.punctuation
+                and tokens[-1] != self._end_of_word_token):
+            tokens.append(self._end_of_word_token)
+        if not list(filter(str.isalpha, text.lower())):
+            return None
+        return tuple(tokens)
+        # if not isinstance(text, (str | None)):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:61]: remove

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -121,6 +179,15 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(encoded_corpus, tuple) or not encoded_corpus:
+            return None
+        decoded_text = self._decode(encoded_corpus)
+        if not decoded_text:
+            return None
+        post_text = self._postprocess_decoded_text(decoded_text)
+        if not post_text:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:188]: `return post_text or None`

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -104,6 +157,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or not len(element) == 1:
+            return None
+        if element not in self._storage:
+            self._storage[element] = len(self._storage)
+        return None
```
[the remark this answers is not in the document]
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py:150]: guess I found it)

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -223,7 +310,7 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
         """
 
     def _extract_n_grams(
-        self, encoded_corpus: tuple[int, ...]
+            self, encoded_corpus: tuple[int, ...]
```
[the remark this answers is not in the document]
Comment (a00rina) [reply lab_3_generate_by_ngrams/main.py]: done

**Pull request closed without merging**